### PR TITLE
Map ItineraryCard to new GraphQL structure

### PR DESCRIPTION
**Pull request by jvaclavik**
**Done**
- Rename `ConnectionCard` to `ItineraryCard`
- Refactor `ItineraryCard` according to new structure
- `ItineraryCard` with collapsed segments
- Add multiple carrier's logos in card
- Fixing types, lint
- Fixing unused fields in relay

### CODE REVIEW

File: packages/components/package.json [hunk 1]
```diff
@@ -6,7 +6,8 @@
   "dependencies": {
     "@kiwicom/margarita-utils": "^0",
     "@kiwicom/universal-components": "0.0.11",
-    "react-native-shimmer-placeholder": "https://github.com/tomzaku/react-native-shimmer-placeholder.git#expo"
+    "react-native-shimmer-placeholder": "https://github.com/tomzaku/react-native-shimmer-placeholder.git#expo",
+    "shortid": "^2.2.14"
```
Comment (RobinCsl) [inline packages/components/package.json]: Why do we need this? Would there be a way to do without this dependency?

File: packages/components/src/ItineraryCard/TripSector.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+const timeSimpleFormat = 'H:mm';
+const dateFormat = 'ddd D MMM';
+
+const getFormattedDate = (time, format = timeSimpleFormat) =>
+  DateFNS.format(DateFNS.parse(time), format);
+
+const getDuration = durationInMinutes => {
+  return (
+    durationInMinutes &&
+    `${Math.floor(durationInMinutes / 60)}h ${durationInMinutes % 60}m`
```
Comment (RobinCsl) [inline packages/components/src/ItineraryCard/TripSector.js]: This is not going to be translation friendly. Can DateFNS deal with it somehow?

File: packages/components/index.js [hunk 3]
```diff
@@ -27,11 +27,7 @@ export {
 export { PlaceCard } from './src/PlaceCard';
 export { BlackToAlpha };
 export { Shimmer } from './src/Shimmer';
-export { ConnectionCard } from './src/ConnectionCard';
-export type {
-  TripSegmentWithId,
-} from './src/ConnectionCard/ConnectionCardTypes';
-
+export { ItineraryCard, ItineraryCardType } from './src/ItineraryCard';
```
Comment (JoseDuSV) [inline packages/components/index.js]: `ItineraryCardType` export can be probably removed, it doesn't seem to be used anywhere and is currently causing warning below.

screenshot 2019-02-07 at 16 20 11

File: packages/components/src/ItineraryCard/TripSector.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  );
+};
+
+const mapTransporters = segments => {
```
Comment (JoseDuSV) [inline packages/components/src/ItineraryCard/TripSector.js]: Can you please move all the helpers and parsers like `mapTransporters`, `getDuration`, `getFormattedDate` into separate file ( for example `TripSectorHelpers`) so I can easily reuse them inside web version.

File: packages/components/src/ItineraryCard/TripSector.js (hunk 2 above)
Comment (jvaclavik) [reply packages/components/src/ItineraryCard/TripSector.js]: True, I will check it

File: packages/components/src/ItineraryCard/TripSector.js (hunk 4 above)
Comment (jvaclavik) [reply packages/components/src/ItineraryCard/TripSector.js]: Good idea. I will do it

File: packages/components/package.json (hunk 1 above)
Comment (FilipMessa) [reply packages/components/package.json]: The current use case of this is just for the key in `map`.
And we should not generate the Id on GraphQL, because it later can lead to misunderstanding.

File: packages/components/index.js (hunk 3 above)
Comment (jvaclavik) [reply packages/components/index.js]: OK, removed

File: scripts/relay-compiler.js [hunk 5]
```diff
@@ -21,7 +21,7 @@ fs.writeFileSync(path.join(__dirname, '..', 'schema.graphql'), clientSchema);
 
 _x('yarn', [
   'relay-compiler',
-  '--src=./apps',
+  '--src=./',
```
Comment (tbergq) [inline scripts/relay-compiler.js]: Why this change?

File: packages/components/src/ItineraryCard/TripSector.js (hunk 2 above)
Comment (jvaclavik) [reply packages/components/src/ItineraryCard/TripSector.js]: Fixed

File: packages/components/src/ItineraryCard/TripSector.js (hunk 4 above)
Comment (jvaclavik) [reply packages/components/src/ItineraryCard/TripSector.js]: Fixed. I also had to create new fragment for `transporters`

File: scripts/relay-compiler.js (hunk 5 above)
Comment (jvaclavik) [reply scripts/relay-compiler.js]: Relic from times when fragment container was in `margarita-components`. 

Fixed :-)

File: apps/core/components/ItineraryCard/ItineraryCard.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      return (
+        <ItineraryCardRow>
+          <TripSector data={item} />
+        </ItineraryCardRow>
+      );
+    }
+    return null;
+  };
+
+  keyExtractor = () => shortid.generate();
```
Comment (tbergq) [inline apps/core/components/ItineraryCard/ItineraryCard.js]: Please use the id field for this purpose.

File: apps/core/components/ItineraryCard/TripSector.js [hunk 7]
```diff
@@ -3,62 +3,52 @@
...
-  carrier,
-  departure,
-  departureTime,
-  duration,
-  tripDate,
-}: Props) {
+function TripSector({ data }: Props) {
+  const firstSegment = data?.segments?.[0];
+  const lastSegment =
+    data && data.segments && data.segments[data.segments.length - 1];
```
Comment (tbergq) [inline apps/core/components/ItineraryCard/TripSector.js]: please use optionalChaining together with last from rambda here. 

```
const segments = data?.segments ?? [];
const firstSegment = first(segments);
const lastSegment = last(segments);
```

File: apps/core/components/ItineraryCard/TripSector.js [hunk 8]
```diff
@@ -67,6 +57,31 @@ export default function TripSegment({
...
+  graphql`
+    fragment TripSector on Sector {
+      duration
+      segments {
+        arrivalTime {
+          local
+        }
+        departureTime {
+          local
+        }
```
Comment (tbergq) [inline apps/core/components/ItineraryCard/TripSector.js]: You could use another fragment for this, so you don't have to repeat this logic

File: apps/core/components/ItineraryCard/TripSectorHelpers.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  time: ?string,
+  format: ?string = timeSimpleFormat,
+) => DateFNS.format(DateFNS.parse(time), format);
+
+export const getDuration = (durationInMinutes: ?number) => {
+  if (!durationInMinutes) {
+    return null;
+  }
+  const date = new Date(0, 0, 0, 0, durationInMinutes);
+  return durationInMinutes && DateFNS.format(date, durationFormat);
```
Comment (tbergq) [inline apps/core/components/ItineraryCard/TripSectorHelpers.js]: do you need durationInMinutoes here, you already checked in the if that it is not falsy

File: apps/core/components/ItineraryCard/ItineraryCard.js (hunk 6 above)
Comment (jvaclavik) [reply apps/core/components/ItineraryCard/ItineraryCard.js]: But server is not getting you any ID of sector

File: packages/components/package.json (hunk 1 above)
Comment (tbergq) [reply packages/components/package.json]: We should generate some ID field for this on the server. We should not use shortID. It beats the purpose of the [key](https://reactjs.org/docs/lists-and-keys.html#keys) prop.

File: apps/core/components/ItineraryCard/ItineraryCard.js (hunk 6 above)
Comment (tbergq) [reply apps/core/components/ItineraryCard/ItineraryCard.js]: Then we should create an ID for sector

Comment (mrtnzlml) [reply apps/core/components/ItineraryCard/ItineraryCard.js]: You should use stable keys for this purpose. It's not complicated to add a new unique and stable ID. You could as well replace this with `keyExtractor = (_, index) => index;` which does the same thing but without any external library.

File: apps/core/components/ItineraryCard/TripSector.js (hunk 7 above)
Comment (jvaclavik) [reply apps/core/components/ItineraryCard/TripSector.js]: Nice, it's much better. Thanks!

File: apps/core/components/ItineraryCard/TripSectorHelpers.js (hunk 9 above)
Comment (jvaclavik) [reply apps/core/components/ItineraryCard/TripSectorHelpers.js]: You are right. Fixed

File: apps/core/components/ItineraryCard/TripSector.js (hunk 8 above)
Comment (jvaclavik) [reply apps/core/components/ItineraryCard/TripSector.js]: Could you explain it more, please? I don't know what exactly you mean.

File: packages/components/package.json (hunk 1 above)
Comment (jvaclavik) [reply packages/components/package.json]: So, ID is now sector index which is get from GraphQL, is it OK?

Comment (tbergq) [reply packages/components/package.json]: Anything that uniquely identifies the sector is useable

File: apps/core/components/ItineraryCard/TripSector.js (hunk 8 above)
Comment (tbergq) [reply apps/core/components/ItineraryCard/TripSector.js]: I can show you on monday 😊

File: apps/graphql/src/dataloaders/itinerariesHelpers.js [hunk 10]
```diff
@@ -61,10 +62,11 @@ export const mapSectors = (
...
 
         // Second map is used for mapping proper GraphQL structure
-        .map(sector => {
-          const firstSegment = sector[0];
-          const lastSegment = sector[sector.length - 1];
+        .map((sector, key) => {
+          const firstSegment = head(sector);
+          const lastSegment = last(sector);
           return {
+            id: `${key}`,
```
Comment (tbergq) [inline apps/graphql/src/dataloaders/itinerariesHelpers.js]: I am not convinced this will be unique.

File: apps/core/components/ItineraryCard/ItineraryCard.js (hunk 6 above)
Comment (FilipMessa) [reply apps/core/components/ItineraryCard/ItineraryCard.js]: So is it OK generate ID on the server if there is no exist ID? 
I think that it can lead to confusion because it creates the expectation that you can ask for a particular sector by ID, but it is no truth.

Comment (mrtnzlml) [reply apps/core/components/ItineraryCard/ItineraryCard.js]: Should be fine, yes. ID doesn't say anything about fetching. It just uniquely represents the type (exactly what you need here). Later you could use it for fetching without BC breaks if there is going to be such a query.

As I said - if you don't want it's fine to use just `keyExtractor = (_, index) => index` - it's better than generating random keys every time. Keep it simple.

Comment (FilipMessa) [reply apps/core/components/ItineraryCard/ItineraryCard.js]: @mrtnzlml 
thanks for the explanation, 👍 make it more sense to me now.
